File: mymail-platform/src/routes/mail.js
const express = require('express');
const multer = require('multer');
const path = require('path');
const fs = require('fs');
const router = express.Router();
const config = require('../config');
const messageDAO = require('../dao/message-dao');
const attachmentDAO = require('../dao/attachment-dao');
const sendLogDAO = require('../dao/send-log-dao');
const userDAO = require('../dao/user-dao');
const { authenticate } = require('../middleware/auth');
const { sendMail } = require('../services/smtp-sender');
const archiver = require('archiver');
const logger = require('../logger');

// Multer config for attachments
const storage = multer.diskStorage({
  destination: (req, file, cb) => {
    const userDir = path.join(config.mail.attachmentPath, String(req.user.id));
    fs.mkdirSync(userDir, { recursive: true });
    cb(null, userDir);
  },
  filename: (req, file, cb) => {
    const unique = Date.now() + '-' + Math.round(Math.random() * 1e9);
    cb(null, unique + '-' + Buffer.from(file.originalname, 'latin1').toString('utf8'));
  },
});

const upload = multer({
  storage,
  limits: { fileSize: config.mail.maxAttachmentSize },
  fileFilter: (req, file, cb) => {
    if (config.allowedAttachmentTypes.includes(file.mimetype)) {
      cb(null, true);
    } else {
      cb(new Error('不支持的文件类型: ' + file.mimetype));
    }
  },
});

router.use(authenticate);

// GET /api/mail/list
router.get('/list', (req, res) => {
  const { folder, page, limit, search, unread } = req.query;
  const result = messageDAO.findByUserId(req.user.id, {
    folder: folder || 'INBOX',
    page: parseInt(page) || 1,
    limit: parseInt(limit) || 20,
    search,
    unreadOnly: unread === 'true',
  });
  res.json({
    ...result,
    unreadCount: messageDAO.unreadCount(req.user.id, folder || 'INBOX'),
  });
});

// GET /api/mail/unread-count
router.get('/unread-count', (req, res) => {
  const folders = ['INBOX', 'SENT', 'DRAFTS', 'TRASH'];
  const counts = {};
  for (const f of folders) {
    counts[f] = messageDAO.unreadCount(req.user.id, f);
  }
  res.json(counts);
});

// GET /api/mail/:id
router.get('/:id', (req, res) => {
  const msg = messageDAO.findById(parseInt(req.params.id));
  if (!msg || msg.user_id !== req.user.id) {
    return res.status(404).json({ error: '邮件不存在' });
  }
  const attachments = attachmentDAO.findByMessageId(msg.id);
  messageDAO.markRead(msg.id);
  res.json({ ...msg, attachments });
});

// POST /api/mail/send
router.post('/send', upload.array('attachments', 10), async (req, res) => {
  try {
    const { to, cc, bcc, subject, bodyHtml, bodyText, replyTo } = req.body;
    if (!to) {
      return res.status(400).json({ error: '收件人不能为空' });
    }

    // Validate subject length
    if (subject && subject.length > 500) {
      return res.status(400).json({ error: '主题最多500字符' });
    }

    // Validate body size (1MB)
    const bodySize = (bodyHtml || '').length + (bodyText || '').length;
    if (bodySize > 1024 * 1024) {
      return res.status(400).json({ error: '正文太大，最多1MB' });
    }

    // Rate limiting
    const recentCount = sendLogDAO.recentCount(req.user.id, 60000);
    if (recentCount >= config.rateLimit.sendPerMin) {
      return res.status(429).json({ error: `发送太频繁，每分钟最多${config.rateLimit.sendPerMin}封` });
    }

    const toAddrs = to.split(',').map(s => s.trim()).filter(Boolean);
    const ccAddrs = cc ? cc.split(',').map(s => s.trim()).filter(Boolean) : [];
    const bccAddrs = bcc ? bcc.split(',').map(s => s.trim()).filter(Boolean) : [];
    const allRecipients = [...toAddrs, ...ccAddrs, ...bccAddrs];

    // Validate email format
    const emailRegex = /^[^\s@]+@[^\s@]+$/;
    for (const addr of allRecipients) {
      if (!emailRegex.test(addr)) {
        return res.status(400).json({ error: `无效的邮箱地址: ${addr}` });
      }
    }

    const files = (req.files || []).map(f => ({
      filename: Buffer.from(f.originalname, 'latin1').toString('utf8'),
      path: f.path,
      size: f.size,
      mimeType: f.mimetype,
    }));

    // Send via SMTP
    const result = await sendMail({
      from: { name: req.user.display_name, address: req.user.email },
      to: toAddrs,
      cc: ccAddrs,
      bcc: bccAddrs,
      subject,
      html: bodyHtml || bodyText,
      text: bodyText,
      replyTo: replyTo || req.user.email,
      attachments: files,
    });

    // Save to SENT
    const totalSize = files.reduce((sum, f) => sum + f.size, 0);
    const msgId = messageDAO.create({
      userId: req.user.id,
      folder: 'SENT',
      messageId: result.messageId,
      uid: messageDAO.getNextUid(req.user.id),
      fromAddr: req.user.email,
      fromName: req.user.display_name,
      toAddr: toAddrs.join(', '),
      ccAddr: ccAddrs.join(', ') || null,
      bccAddr: bccAddrs.join(', ') || null,
      replyTo: replyTo || req.user.email,
      subject,
      bodyText: bodyText || '',
      bodyHtml: bodyHtml || '',
      hasAttach: files.length > 0,
      attachCount: files.length,
      sizeBytes: totalSize,
    });

    // Save attachment records
    for (const file of files) {
      attachmentDAO.create({
        messageId: msgId,
        filename: file.filename,
        mimeType: file.mimeType,
        sizeBytes: file.size,
        storagePath: file.path,
      });
    }

    // Log
    sendLogDAO.create({ userId: req.user.id, toAddr: toAddrs.join(', '), subject, status: 'sent' });
    userDAO.updateStorageUsed(req.user.id, totalSize);

    res.json({ message: '发送成功', messageId: result.messageId });
  } catch (err) {
    logger.error({ err }, 'Send error');
    sendLogDAO.create({ userId: req.user.id, toAddr: req.body.to, subject: req.body.subject, status: 'failed', errorMsg: err.message });
    res.status(500).json({ error: '发送失败: ' + err.message });
  }
});

// POST /api/mail/save-draft
router.post('/save-draft', (req, res) => {
  const { to, cc, bcc, subject, bodyHtml, bodyText } = req.body;
  const msgId = messageDAO.create({
    userId: req.user.id,
    folder: 'DRAFTS',
    messageId: `draft-${Date.now()}@${config.domain}`,
    uid: messageDAO.getNextUid(req.user.id),
    fromAddr: req.user.email,
    fromName: req.user.display_name,
    toAddr: to || '',
    ccAddr: cc || null,
    bccAddr: bcc || null,
    subject: subject || '(无主题)',
    bodyText: bodyText || '',
    bodyHtml: bodyHtml || '',
  });
  res.json({ message: '草稿已保存', id: msgId });
});

// PUT /api/mail/:id/read
router.put('/:id/read', (req, res) => {
  messageDAO.markRead(parseInt(req.params.id));
  res.json({ message: 'ok' });
});

// PUT /api/mail/:id/unread
router.put('/:id/unread', (req, res) => {
  messageDAO.markUnread(parseInt(req.params.id));
  res.json({ message: 'ok' });
});

// PUT /api/mail/:id/star
router.put('/:id/star', (req, res) => {
  messageDAO.toggleStar(parseInt(req.params.id));
  res.json({ message: 'ok' });
});

// DELETE /api/mail/:id
router.delete('/:id', (req, res) => {
  const msg = messageDAO.findById(parseInt(req.params.id));
  if (!msg || msg.user_id !== req.user.id) return res.status(404).json({ error: '邮件不存在' });
  if (msg.folder === 'TRASH') {
    messageDAO.permanentDelete(msg.id);
  } else {
    messageDAO.moveToFolder(msg.id, 'TRASH');
  }
  res.json({ message: 'ok' });
});

// PUT /api/mail/:id/restore
router.put('/:id/restore', (req, res) => {
  const msg = messageDAO.findById(parseInt(req.params.id));
  if (!msg || msg.user_id !== req.user.id) return res.status(404).json({ error: '邮件不存在' });
  const folder = msg.in_reply_to ? 'INBOX' : 'INBOX';
  messageDAO.moveToFolder(msg.id, 'INBOX');
  res.json({ message: '已恢复' });
});

// POST /api/mail/empty-trash
router.post('/empty-trash', (req, res) => {
  messageDAO.emptyTrash(req.user.id);
  res.json({ message: '垃圾箱已清空' });
});

// GET /api/mail/:id/attachments/:aid/download
router.get('/:id/attachments/:aid/download', (req, res) => {
  const att = attachmentDAO.findById(parseInt(req.params.aid));
  if (!att) return res.status(404).json({ error: '附件不存在' });
  const msg = messageDAO.findById(att.message_id);
  if (!msg || msg.user_id !== req.user.id) return res.status(403).json({ error: '无权访问' });
  res.download(att.storage_path, att.filename);
});

// GET /api/mail/:id/attachments/download-all
router.get('/:id/attachments/download-all', (req, res) => {
  const msgId = parseInt(req.params.id);
  const msg = messageDAO.findById(msgId);
  if (!msg || msg.user_id !== req.user.id) return res.status(403).json({ error: '无权访问' });
  const attachments = attachmentDAO.findByMessageId(msgId);
  if (!attachments.length) return res.status(404).json({ error: '没有附件' });

  res.setHeader('Content-Type', 'application/zip');
  res.setHeader('Content-Disposition', 'attachment; filename="attachments.zip"');
  
  const archive = archiver('zip', { zlib: { level: 9 } });
  archive.pipe(res);
  for (const att of attachments) {
    archive.file(att.storage_path, { name: att.filename });
  }
  archive.finalize();
});

module.exports = router;

File: mymail-go/internal/httpapi/dto/mail.go
// Package dto
// mail.go 定义邮件相关 HTTP 请求/响应 DTO。
//
// 字段命名约定（与原 Node.js 后端 100% 兼容）：
//   - 请求 body：camelCase（前端 ComposeView.vue 用 bodyHtml/bodyText/replyTo）
//   - 响应 JSON：messages 表字段 snake_case（前端 MailView.vue 用 mail.is_read/mail.from_addr）
//   - 响应包装字段：camelCase（unreadCount/messageId）
package dto

// ============ 请求 DTO ============

// SaveDraftRequest 保存草稿请求（JSON body）。
// 字段为 camelCase，与前端一致。
type SaveDraftRequest struct {
	To       string `json:"to"`
	Cc       string `json:"cc"`
	Bcc      string `json:"bcc"`
	Subject  string `json:"subject"`
	BodyHTML string `json:"bodyHtml"`
	BodyText string `json:"bodyText"`
}

// BatchOperationRequest 批量操作请求（JSON body）。
// 用于 batch/mark-read、batch/move、batch/delete。
// IDs 为邮件 ID 列表；Folder 仅 batch/move 用。
type BatchOperationRequest struct {
	IDs    []int64 `json:"ids" binding:"required"`
	Folder string  `json:"folder,omitempty"`
}

// ============ 响应 DTO ============

// MailDetailResponse 邮件详情响应（snake_case，与原 Node.js SELECT * 一致）。
// 前端 MailDetailView.vue 直接消费这些字段名。
// 命名为 MailDetailResponse 避免与 dto/auth.go 的通用 MessageResponse 冲突。
type MailDetailResponse struct {
	ID            int64                 `json:"id"`
	UserID        int64                 `json:"user_id"`
	Folder        string                `json:"folder"`
	MessageID     string                `json:"message_id"`
	UID           *int64                `json:"uid"`
	FromAddr      string                `json:"from_addr"`
	FromName      string                `json:"from_name"`
	FromAvatarURL string                `json:"from_avatar_url"`
	ToAddr        string                `json:"to_addr"`
	CcAddr      string                `json:"cc_addr"`
	BccAddr     string                `json:"bcc_addr"`
	ReplyTo     string                `json:"reply_to"`
	Subject     string                `json:"subject"`
	BodyText    string                `json:"body_text"`
	BodyHTML    string                `json:"body_html"`
	IsRead      bool                  `json:"is_read"`
	IsStarred   bool                  `json:"is_starred"`
	IsDeleted   bool                  `json:"is_deleted"`
	HasAttach   bool                  `json:"has_attach"`
	AttachCount int                   `json:"attach_count"`
	SizeBytes   int64                 `json:"size_bytes"`
	InReplyTo   string                `json:"in_reply_to,omitempty"`
	Flags       string                `json:"flags"`
	SpamScore   int                   `json:"spam_score"`
	SpamReasons string                `json:"spam_reasons,omitempty"`
	ReceivedAt  string                `json:"received_at"`
	Attachments []*AttachmentResponse `json:"attachments,omitempty"`
}

// AttachmentResponse 附件响应（snake_case）。
// 注意：storage_path 不暴露给前端（安全考虑）。
type AttachmentResponse struct {
	ID        int64  `json:"id"`
	MessageID int64  `json:"message_id"`
	Filename  string `json:"filename"`
	MimeType  string `json:"mime_type"`
	SizeBytes int64  `json:"size_bytes"`
	CreatedAt string `json:"created_at"`
}

// ListResponse 邮件列表响应。
// 包装字段 unreadCount 为 camelCase（与原 Node.js 一致）。
type ListResponse struct {
	Messages    []*MailDetailResponse `json:"messages"`
	Total       int64                 `json:"total"`
	Page        int                   `json:"page"`
	Limit       int                   `json:"limit"`
	UnreadCount int64                 `json:"unreadCount"`
}

// UnreadCountResponse 各文件夹未读数响应。
type UnreadCountResponse struct {
	INBOX  int64 `json:"INBOX"`
	SENT   int64 `json:"SENT"`
	DRAFTS int64 `json:"DRAFTS"`
	TRASH  int64 `json:"TRASH"`
}

// SendResponse 发送邮件响应。
type SendResponse struct {
	Message   string `json:"message"`
	MessageID string `json:"messageId"`
}

// SaveDraftResponse 保存草稿响应。
type SaveDraftResponse struct {
	Message string `json:"message"`
	ID      int64  `json:"id"`
}

// BatchOperationResponse 批量操作响应。
type BatchOperationResponse struct {
	Message  string `json:"message"`
	Affected int64  `json:"affected"`
}

// UploadResponse 上传附件响应（用于 upload 端点）。
type UploadResponse struct {
	Message string             `json:"message"`
	File    *UploadedFileDTO   `json:"file,omitempty"`
}

// UploadedFileDTO 已上传文件元信息。
type UploadedFileDTO struct {
	Filename  string `json:"filename"`
	MimeType  string `json:"mime_type"`
	SizeBytes int64  `json:"size_bytes"`
	Path      string `json:"path"` // 相对路径，供前端引用
}
